tests/hypothesis_strategies/constraints.py

```diff
@@ -0,0 +1,149 @@
+"""Hypothesis strategies for constraints."""
+
+from functools import partial
+from typing import Any, List, Optional, Type, Union
+
+import hypothesis.strategies as st
+
+from baybe.constraints.conditions import (
+    SubSelectionCondition,
+    ThresholdCondition,
+    _valid_logic_combiners,
+)
+from baybe.constraints.continuous import (
+    ContinuousLinearEqualityConstraint,
+    ContinuousLinearInequalityConstraint,
+)
+from baybe.constraints.discrete import (
+    DiscreteExcludeConstraint,
+    DiscreteLinkedParametersConstraint,
+    DiscreteNoLabelDuplicatesConstraint,
+    DiscreteProductConstraint,
+    DiscreteSumConstraint,
+)
+from baybe.parameters.base import DiscreteParameter
+from baybe.parameters.numerical import NumericalDiscreteParameter
+
+
+def sub_selection_conditions(superset: Optional[List[Any]] = None):
+    """Generate :class:`baybe.constraints.conditions.SubSelectionCondition`."""
+    if superset is None:
```

Comment:
Don't quite get the use case for when the superset is `None`, could you elaborate?

Reply:
This is not fully set in stone, but so far I tried to keep the strategies somewhat general so that they can be used flexibly. Consider two use cases:
1. For simple tests (e.g. serialization), you don't actually care what are the available choices. You just want to get a `SubSelectionConditition` object to work with, e.g. to serialize and whatsoever (this particular example is not the best because there is not much to do with that object, but you get the gist ... with other object, like search spaces, you can do a lot). Hence, you want to be able to create that strategy without having to think about fake content.
2. However, there are tests where you want to actually *use* that object for certain parameters, e.g. to test whether certain constraints work. Hence, you need the option to set the specific content. 

Makes sense?

Reply:
thanks for the question and explanation, i also needed that

such things would be welcome as short comment in the code itself

Reply:
One question related to the optional parameters list: in case its None its clear, its a situation where you dont want to create an actual object to do stuff with.

But im wondering whether we should adjust the case where its not None, ie we want a valid object we can do stuff with. Right now the parameters arg is a list of parameters that are affected by the constraint created. But we could also make the choice to embedd in this list the larger context of *all* available parameters and its the responsibility of the constraint to subselect some of them that are affected by it. This makes life a bit easier for the more complex constraints as they might benefit from a larger context. 

Example: If i want to make a dependencies constraint I would select from the list of all parameters N candidates. But then for each candidate I need to select from the overall list M_i other so called affected parameters (ie another list and overall Id have a list of lists), but the list of M_i cannot include the candidate itself (and maybe should also not include any of the other candidates)

i) maybe thats also doable with the current version and Im overthinking it? If I remember correctly theres filter stuff in hypothesis maybe that would already suffice?
ii) maybe the argument is valid but should only be implemented for the complex constraints that need it? ie theyd have an interfacte to `all_parameters` instead of `parameters` or so
iii) maybe for consistency the interface for all constraint strategies should change to reflect that, ie the parameters list is not the list of affected parameters but the list of all existing possible parameters

Reply:
Ive pushed a commit with a proposed (not yet fully elegantized) version for the dependency cosntraint that might illustrate the question / idea

Reply:
I have to admit that I neither understand your comment not your example completely :/ It seems to me like you are proposing to alter the way the Constraint works in general? Or is it just about the test?

Reply:
My quesstion has nothing to do with altering the actual constraints but the hypothesis strategies this PR is concerned with. 

The currently implemented constraint strategies have one list of parameters as (optional) input. But other constraints have more arguments relating to paramter lists and they are not completely independent. So my question is about these two approaches:

1) Alter constraint strategies so that they get a list of ALL campaign parameters and not just the ones they are supposed to affect. Selecting which of the parameters they affect would then be the strategies responsibility.
2) Make it the responsibility of the caller of the hypothesis strategy to provide correct arguments.This would mean for the more complex constraints that they would get additional arguments.

2 Is implicitly implemented as right now. Personally Id find 1 a bit nicer as the responsibility of selecting plausible parameters and combinations thereof is encapsulated in the object. For the simple constraints implemented right now the choice here doesnt really matter but Im asking for the ones we still lack strategies for.

Reply:
Ahhhh now I get it, thanks for the explanation :) Yes, I also think that 1. is the better option, and I think your commit shows quite nicely how to implement this.

Reply:
ive reverted my opinion on this, the new strats now accepts lists of both the parameters and the ones they depend on. Depending on whether they are not None different tuff is done trying to respect all sorts of requirements

I fully expect that some stuff here is not yet fully working in the context of actual search spaces. Eg the conditions also dont repsect reasonable parameter values, which could lead to completely empty searchspaces (see other comment).

This debate however is resolved imo unless in reviewing my added code you want to reopen it